File: post_3/serving/fastapi-onnx/app/model/inputs.py
import numpy as np

from app.types import AdultIncomeInput


def prepare_inputs(data: list[AdultIncomeInput]) -> dict:
    input_dict = {
        "age": np.array([[d.age] for d in data], dtype=np.float32),
        "workclass": np.array([[d.workclass] for d in data], dtype=object),
        "fnlwgt": np.array([[d.fnlwgt] for d in data], dtype=np.float32),
        "education": np.array([[d.education] for d in data], dtype=object),
        "educational_num": np.array(
            [[d.education_num] for d in data], dtype=np.float32
        ),
        "marital_status": np.array([[d.marital_status] for d in data], dtype=object),
        "occupation": np.array([[d.occupation] for d in data], dtype=object),
        "relationship": np.array([[d.relationship] for d in data], dtype=object),
        "race": np.array([[d.race] for d in data], dtype=object),
        "gender": np.array([[d.sex] for d in data], dtype=object),
        "capital_gain": np.array([[d.capital_gain] for d in data], dtype=np.float32),
        "capital_loss": np.array([[d.capital_loss] for d in data], dtype=np.float32),
        "hours_per_week": np.array(
            [[d.hours_per_week] for d in data], dtype=np.float32
        ),
        "native_country": np.array([[d.native_country] for d in data], dtype=object),
    }

    return input_dict

File: post_3/serving/axum-onnx/src/model/inputs.rs
use crate::types::AdultIncomeInput;
use ndarray::Array;
use ort::value::Value;

pub fn prepare_inputs(data: &[AdultIncomeInput]) -> Result<Vec<Value>, String> {
    use ort::value::Tensor;
    let batch_size = data.len();

    let make_string_array = |v: Vec<String>| {
        Tensor::from_string_array(([batch_size, 1], &v[..]))
            .map_err(|e| format!("Failed to create string tensor: {}", e))
            .map(|t| t.into())
    };

    let make_float_array = |values: Vec<f32>| {
        Array::from_shape_vec((batch_size, 1), values)
            .map_err(|e| format!("Failed to create array: {}", e))
            .and_then(|arr| {
                Value::from_array(arr)
                    .map_err(|e| format!("Failed to create value: {}", e))
                    .map(|v| v.into())
            })
    };

    Ok(vec![
        make_float_array(data.iter().map(|d| d.age).collect())?,
        make_string_array(data.iter().map(|d| d.workclass.clone()).collect())?,
        make_float_array(data.iter().map(|d| d.fnlwgt).collect())?,
        make_string_array(data.iter().map(|d| d.education.clone()).collect())?,
        make_float_array(data.iter().map(|d| d.education_num).collect())?,
        make_string_array(data.iter().map(|d| d.marital_status.clone()).collect())?,
        make_string_array(data.iter().map(|d| d.occupation.clone()).collect())?,
        make_string_array(data.iter().map(|d| d.relationship.clone()).collect())?,
        make_string_array(data.iter().map(|d| d.race.clone()).collect())?,
        make_string_array(data.iter().map(|d| d.sex.clone()).collect())?,
        make_float_array(data.iter().map(|d| d.capital_gain).collect())?,
        make_float_array(data.iter().map(|d| d.capital_loss).collect())?,
        make_float_array(data.iter().map(|d| d.hours_per_week).collect())?,
        make_string_array(data.iter().map(|d| d.native_country.clone()).collect())?,
    ])
}
